### DISCUSSION TITLE
Debugging tips for stm32/gap8 debugging

### DISCUSSION THREAD

**Discussion by finnfi** [bitcraze/discussions#108] [Q&A] [answered]
Hi,

I am relatively new to the multiple mcu debugging field. 

I am working on an app on the stm32 and some custom code for the GAP8. 

I've read the debugging guide for the stm32 [https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/](url). Does something like this exist for the gap8 as well? Since I am using the docker container (aideck-with-autotiler) to build, I am not sure how to use a debugger. 

Another simpler solution would be to get console output (printf) from both mcus to the desktop screen. 

How do you usually do it? 

Finn

**Comment by evoggy** [comment, marked as answer]
Hi!

Yeah, developing for multi-mcu systems can be a hassle :grimacing: When I've been developing for the AI deck I've been using the following setup:
- SWD dongle (ST Link V2) for the STM32
- UART printout from the ESP32 (on IO1 on the expansion port). **NOTE** This might interfere with the bootloading since you're hijacking the boot pin and using it for UART output, instead I've been using the Olimex JTAG for programming.
- UART printout from the GAP8 (RX1 on the expansion port)

Here's [a link to the schematic](https://www.bitcraze.io/documentation/hardware/ai_deck_1_1/ai-deck-revc.pdf) to see the pins.

This becomes a complex (and large) setup with lots of adapters. Since I've been working on the CPX communication I haven't been able to use that channel for printing. Another option for you would be to use the printouts over CPX/CRTP, which would mean you don't need the adapters. In the Crazyflie firmware you can use ```DEBUG_PRINT``` and in the GAP8 you can use ```cpxPrintToConsole``` like this:
```c
/**
 * @brief Print debug data though CPX
 *
 * This will print debug data though CPX to the Crazyflie client console.
 * The function doesn't add a newline, so this will have to be supplied.
 *
 * @param target The target where the printout should go
 * @param fmt Standard C format string
 * @param variable Data for format string
 */
void cpxPrintToConsole(CPXConsoleTarget_t target, const char * fmt, ...);
```

Then all the printouts will be shown together in the console in the Crazyflie python client. But note that these are asynchronous, so if you for instance crash somewhere you might not get the printouts out before the crash.

**Reply by finnfi** [reply]
Thanks!
